src/git/SourceControlPanel.cpp

```diff
@@ -0,0 +1,672 @@
+
+	auto repositoryLayout = BLayoutBuilder::Group<>(B_VERTICAL, B_USE_BORDER_INSETS)
+		.SetInsets(0)
+		.AddGroup(B_VERTICAL)
+			.SetInsets(5)
+			.Add(fBranchMenu)
+			.End()
+		.Add(fToolBar)
+		.Add(fPanelsLayout)
+		.View();
+
+	fMainLayout = BLayoutBuilder::Cards<>()
+		.Add(repositoryLayout)
+		.Add(fRepositoryNotInitializedView)
+		.SetVisibleItem((int32)kMainIndexRepository);
+
+	fToolBar->ToggleActionPressed(MsgShowRepositoryPanel);
+
+	BLayoutBuilder::Group<>(this, B_VERTICAL, B_USE_BORDER_INSETS)
+		.SetInsets(0)
+		.AddGroup(B_VERTICAL)
+			.SetInsets(5)
+			.Add(fProjectMenu)
+			.End()
+		.Add(fMainLayout)
+		.End();
+}
+
+
+SourceControlPanel::~SourceControlPanel()
+{
+	delete fLogView;
+	delete fChangesView;
+	delete fRepositoryView;
+	delete fToolBar;
+	delete fBranchMenu;
+	delete fProjectMenu;
+	delete fRepositoryNotInitializedView;
+}
+
+
+void
+SourceControlPanel::_InitToolBar()
+{
+	fToolBar = new ToolBar();
+	fToolBar->ChangeIconSize(16);
+	fToolBar->AddAction(MsgShowRepositoryPanel, B_TRANSLATE("Repository"), "kIconGitRepo");
+	// fToolBar->AddAction(MsgShowChangesPanel, B_TRANSLATE("Changes"), "kIconGitChanges");
+	// fToolBar->AddAction(MsgShowLogPanel, B_TRANSLATE("Log"), "kIconGitLog");
+	fToolBar->AddGlue();
+	fToolBar->AddAction(MsgShowOptionsMenu, B_TRANSLATE("Options"), "kIconGitMore");
+}
+
+
+void
+SourceControlPanel::_InitRepositoryView()
+{
+	fRepositoryView = new RepositoryView();
+	fRepositoryViewScroll = new BScrollView(B_TRANSLATE("Repository scroll view"),
+		fRepositoryView, B_FRAME_EVENTS | B_WILL_DRAW, true, true, border_style::B_NO_BORDER);
+}
+
+
+void
+SourceControlPanel::_InitChangesView()
+{
+	fChangesView = BLayoutBuilder::Group<>(B_HORIZONTAL, B_USE_BORDER_INSETS)
+		.AddGroup(B_VERTICAL)
+			.AddGlue()
+			.Add(new BButton("changes"))
+			.AddGlue()
+		.End()
+		.AddGlue()
+		.View();
+}
+
+
+void
+SourceControlPanel::_InitLogView()
+{
+	fLogView = BLayoutBuilder::Group<>(B_HORIZONTAL, B_USE_BORDER_INSETS)
+		.AddGroup(B_VERTICAL)
+			.AddGlue()
+			.Add(new BButton("log"))
+			.AddGlue()
+		.End()
+		.AddGlue()
+		.View();
+}
+
+
+void
+SourceControlPanel::_InitRepositoryNotInitializedView()
+{
+	auto stringView = new BStringView("InitMessage",
+		B_TRANSLATE("This project does not have a git repository.\n"
+			" Click below to initialize it"));
+	fInitializeButton = new BButton(B_TRANSLATE("Init repository"),
+		new BMessage(MsgInitializeRepository));
+	fDoNotCreateInitialCommitCheckBox = new BCheckBox(B_TRANSLATE("Do not create the initial commit"),
+		nullptr);
+	stringView->SetAlignment(B_ALIGN_CENTER);
+	fRepositoryNotInitializedView = BLayoutBuilder::Group<>(B_VERTICAL, B_USE_BORDER_INSETS)
+		.AddGroup(B_VERTICAL)
+			.SetInsets(20)
+			.AddGlue()
+			.Add(stringView)
+			.Add(fInitializeButton)
+			.Add(fDoNotCreateInitialCommitCheckBox)
+			.AddGlue()
+		.End()
+		.AddGlue()
+		.View();
+}
+
+
+void
+SourceControlPanel::AttachedToWindow()
+{
+	if (gMainWindow->Lock()) {
```

Comment:
why not Looper() which is more generic?

Reply:
I had a look at other code (mine included) and here usually we use "LockLooper()". Could you try? I suspect the Window() pointer could be already invalid in certain cases at this point. LockLooper () would fail in that case, and it's more correct

Reply:
I think I've tried at some point this as well to no avail. I'll try again, though. The thing is that the version with Window() used to work before the StartWatching on the ProjectBrowser

Reply:
That would point at a deeper problem, though

Reply:
To be clear, I meant LockLooper/Unlocklooper, not Looper()->Lock()

Reply:
I've tried all possible combinations and they all crash Genio at exit:

Window()->LockLooper() / Window()->UnlockLooper()
gMainWindow->LockLooper() / gMainWindow->UnlockLooper()
Window()->Lock() / Window()->Unock()
Looper()->Lock() / Looper()->Unock()

The only one that works is gMainWindow->Lock() / gMainWindow->Unlock() as long as we watch ProjectBrowser

Reply:
Except the one I suggested last (LockLooper () directly on the View (without passing through the Window pointer)) 😅

Reply:
[Genio_debug-21474-debug-02-12-2023-23-39-44.report.txt](https://github.com/Genio-The-Haiku-IDE/Genio/files/13539020/Genio_debug-21474-debug-02-12-2023-23-39-44.report.txt)

Reply:
gMainWindow seems to be nullptr at this point. Using an explicit cast from Window() to GenioWindow semmes to fix the weird behaviour